Introduce getFieldTitle as field / series display identity  and use it in all field name matchers & field / series name displays

ryantxu opened the pull request:
This PR tries to add true label support to Grafana. That is, make  Grafana really use labels in series (field) identity and display. 

This is relevant to do know as doing it later has an impact on transformations that refer to fields and override rule matches as well. 

So with DataFrames, that have an optional name, we have fields (that have name, usually 'Value' for time series), and fields can have labels. So the question is how to refer to a field and show a field. 

This PR introduces a field.state object that includes a title property that is a dynamic identity for the field and is dependant on what other data is available. Optimized to include as little duplicate information as possible (optimal for display). 

Examples:
Query returns 2 series , but the same metric name, and only one label name, series name (field title) is then only shown using the label value.

`topk(5, max(scrape_duration_seconds) by (job))` 

Returns no name label but but only one label name (job)

![Screenshot from 2020-05-04 19-21-40](https://user-images.githubusercontent.com/10999/80994143-91971100-8e3c-11ea-8e5e-14a6fc428905.png)

If two label names are return they are of course rendered as usual: 
![Screenshot from 2020-05-04 19-24-44](https://user-images.githubusercontent.com/10999/80994430-f3577b00-8e3c-11ea-9e61-a6d3abbc263a.png)

But if all series return the same metric name (__name__) label we omit that (as it woudl be duplicated on all). And is probably something you put in the panel title (if the graph is only showing one metric). 

Makes Grafana automatically shows only relevant information automatically and users will no longer have to use legend formats to get simple short legend names. 

TODO:
- [x] Fix series name for legacy format 
- [ ] Support metrictank tags
- [x] Fix a bunch of tests
- [x] Update transformer matchers to use state.title 

## breaking change

Promethus and graphite is returning time series with tags/labels but with title set, disabling automatic series naming. So there should be no breaking behavior for those data sources.

ryantxu committed:
common title handling

ryantxu committed:
Merge remote-tracking branch 'origin/master' into better-label-support

ryantxu committed:
show labels

ryantxu commented on packages/grafana-data/src/field/fieldDisplay.ts:
```diff
@@ -36,6 +38,7 @@ export interface ReduceDataOptions {
 export const VAR_FIELD_NAME = '__field.name';
+export const VAR_FIELD_LABELS = '__field.labels';
 export const VAR_CALC = '__calc';
 export const VAR_CELL_PREFIX = '__cell_'; // consistent with existing table templates
```
maybe unrelated... but we should remove this __cell_N syntax since it will explode.

We could add:
```
__data.row[N].fieldA.text
```
to give you access to anything in the response.  DataLinks already use __data though, so that may break them or be inconsistent

torkelo replied:
yea, really need to change this one.

ryantxu commented on packages/grafana-data/src/field/fieldDisplay.ts:
```diff
@@ -65,10 +68,33 @@ function getTitleTemplate(title: string | undefined, stats: string[], data?: Dat
   if (fieldCount > 1 || !parts.length) {
     parts.push('${' + VAR_FIELD_NAME + '}');
   }
+  parts.push('${' + VAR_FIELD_LABELS + '}');
```
will return empty when no labels exist

ryantxu committed:
update comment

marefr committed:
Update changelog for v7.0.0-beta1 (#24007)

Co-Authored-By: Marcus Efraimsson <marcus.efraimsson@gmail.com>
Co-Authored-By: Andrej Ocenas <mr.ocenas@gmail.com>
Co-Authored-By: Hugo Häggmark <hugo.haggmark@grafana.com>
Co-authored-by: Tobias Skarhed <1438972+tskarhed@users.noreply.github.com>

aknuds1 committed:
verify-repo-update: Fix Dockerfile.deb (#24030)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
CircleCI: Upgrade build pipeline tool (#24021)

* CircleCI: Upgrade build pipeline tool

xlson committed:
Devenv: ignore enterprise (#24037)

ivanahuckova committed:
Add header icon to Add data source page (#24033)

aknuds1 committed:
latest.json: Update testing version (#24038)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

alexanderzobnin committed:
Fix login page redirected from password reset (#24032)

tskarhed committed:
Storybook: Rewrite stories to CSF (#23989)

* ColorPicker to CSF format

* Convert stories to CSF

* Do not export ClipboardButton

* Update ConfirmButton

* Remove unused imports

* Fix feedback

xlson committed:
changelog enterprise 7.0.0-beta1 (#24039)

aknuds1 committed:
CircleCI: Bump grafana/build-container revision (#24043)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

torkelo committed:
Changelog: Updates changelog with more feature details (#24040)

* Changelog: Updates changelog with more feature details

* spell fix

* spell fix

* Updates

* Readme update

* Updates

mckn committed:
Select: fixes so component loses focus on selecting value or pressing outside of input. (#24008)

* changed the value container to a class component to get it to work with focus (maybe something with context?).

* added e2e tests to verify that the select focus is working as it should.

* fixed according to feedback.

* updated snapshot.

xlson committed:
Devenv: add remote renderer to grafana (#24050)

dprokop committed:
NewPanelEditor: minor UI twekas (#24042)

* Forward ref for tabs, use html props

* Inspect:  add inspect label to drawer title

* Add tooltips to sidebar pane tabs, copy changes

* Remove unused import

* Place tooltips over tabs

* Inspector: dont show transformations select if there is only one data frame

* Review

aknuds1 committed:
Changelog: Add a breaking change (#24051)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
CircleCI: Unpin grafana/docs-base (#24054)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

Clarity-89 committed:
Search: close overlay on Esc press (#24003)

* Search: Close on Esc

* Search: Increase bottom padding for the last item in section

* Search: Move closing search to keybindingsSrv

* Search: Fix folder view

* Search: Do not move folders if already in folder

torkelo committed:
Docs: Adds deprecation notice to changelog and docs for scripted dashboards (#24060)

RichiH committed:
Update CHANGELOG.md (#24047)

Fix typo

Co-authored-by: Daniel Lee <dan.limerick@gmail.com>

joe-elliott committed:
Documentation: Alternative Team Sync Wording (#23960)

* Alternative wording for team sync docs

Signed-off-by: Joe Elliott <number101010@gmail.com>

* Update docs/sources/auth/team-sync.md

Co-Authored-By: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

mjtrangoni committed:
Fix misspell issues (#23905)

* Fix misspell issues

See,
$ golangci-lint run --timeout 10m --disable-all -E misspell ./...

Signed-off-by: Mario Trangoni <mjtrangoni@gmail.com>

* Fix codespell issues

See,
$ codespell -S './.git*' -L 'uint,thru,pres,unknwon,serie,referer,uptodate,durationm'

Signed-off-by: Mario Trangoni <mjtrangoni@gmail.com>

* ci please?

* non-empty commit - ci?

* Trigger build

Co-authored-by: bergquist <carl.bergquist@gmail.com>
Co-authored-by: Kyle Brandt <kyle@grafana.com>

ryantxu committed:
Merge remote-tracking branch 'origin/master' into better-label-support

ryantxu committed:
fix compile error

ryantxu committed:
better series display

ryantxu committed:
better display

ryantxu committed:
now with prometheus and loki

ryantxu commented on packages/grafana-data/src/dataframe/processDataFrame.ts:
```diff
@@ -84,7 +84,6 @@ function convertTimeSeriesToDataFrame(timeSeries: TimeSeries): DataFrame {
   ];
 
   return {
-    name: timeSeries.target,
```
since target is already the field name, do not duplicate it here

ryantxu committed:
a few more tests

ryantxu commented:
@aocenas @torkelo -- can you take a look at this PR in action and see if anything looks really bad?  The goal is a consistent name+labels strategy for everything.  Prometheus has used:
```
name: abc{a="b"}
labels: {a:"b", __name__:"abc"}
```
And this changes that to:
```
name: abc
labels: {a:"b"}
```

For our usage in grafana, we will need to call `getFieldDisplayTitle()` rather than just field.name... well actually we often call `field.config?.title ?? field.name`!


TODO: metrictank label support

torkelo commented on packages/grafana-data/src/field/fieldDisplay.ts:
```diff
@@ -65,10 +68,69 @@ function getTitleTemplate(title: string | undefined, stats: string[], data?: Dat
+  let name = field.config?.title ?? field.name;
+
+  if (field.labels) {
+    name += formatLabels(field.labels);
```
should we add whitespace between name and labels?

torkelo commented on packages/grafana-data/src/field/fieldDisplay.ts:
```diff
@@ -65,10 +68,69 @@ function getTitleTemplate(title: string | undefined, stats: string[], data?: Dat
+/**
+ * Get an appropriate display title.  If the 'title' is set, use that
+ */
+export function getFieldDisplayTitle(field: Field, frame?: DataFrame) {
+  let title = field.config?.title;
+  if (title) {
+    return title; // Title is set and not a template
+  }
+
+  const id = getFieldId(field);
+  const seriesName = frame?.name;
+  if (seriesName && name !== seriesName) {
+    return seriesName + ' ' + name;
+  }
+  return id;
+}
```
The biggest problem here is that we do not have access to other data frames here to dedupe and remove frame name when the frame name is all the same.

Currently, we have a big breaking behavior for Prometheus still. A query like `sum by(mode)(irate(node_cpu_seconds_total{job="node_exporter", instance=~".*", mode!="idle"}[5m])) > 0`  Returns no __name__ label (because of the sum). So with your updated code in Prometheus result_transformer the frame names are all the same (the query text). So the series names are all:

![Screenshot from 2020-04-30 09-37-57](https://user-images.githubusercontent.com/10999/80684688-55735180-8ac6-11ea-9651-15d15641f9b6.png)

So the frame name is all the same for all the series. It would be easy to fix the logic in the prometheus result transformer to not set frame name to options.query if there are any labels. So, in that case, we can have an empty frame name. Hm.. maybe that is correct. But then it would be impossible to separate two queries that return the same label sets. But maybe not a common/realistic scenario.

torkelo replied:
So to add, that query today (in 6.7) would result in a legend with only {mode="value"} ie only the label key-value pair.

torkelo replied:
I changed the result_transformer in prometheus to set name to undefined if there are other labels (but no __name__) label. But then I get

Value prefix, and here is the problem of not knowing the names of the fields in the other data frames in order to have smart title info removal for duplicated (same) names  :) 
![Screenshot from 2020-04-30 09-47-05](https://user-images.githubusercontent.com/10999/80685541-a768a700-8ac7-11ea-9885-4305c9899034.png)

torkelo reviewed (commented):
This is looking better and better. 

now the labels to fields even work with different result sets (resulting in a more than 1 value column, automatically getting the __name__ label value). 
![Screenshot from 2020-04-30 10-21-57](https://user-images.githubusercontent.com/10999/80688696-7f2f7700-8acc-11ea-9a61-028e1c64a63a.png)

And if I put in legend alias I get __name__ column and custom name for the value column!

torkelo committed:
Improvements and tests

aocenas commented:
Not sure I get all the implications but it does not seem like anything is blowing up at the moment.

torkelo commented on public/app/plugins/datasource/prometheus/result_transformer.ts:
```diff
@@ -141,23 +139,29 @@ export class ResultTransformer {
-      label = this.getOriginalMetricName(labelData);
-    } else {
-      label = this.renderTemplate(this.templateSrv.replace(options.legendFormat), labelData);
+  createLabelInfo(labelData: { [key: string]: string }, options: any): { name: string; labels: Labels } {
+    if (options?.legendFormat) {
+      // Alternativly we could put the results in "title"
+      const name = this.renderTemplate(this.templateSrv.replace(options.legendFormat), labelData);
+      if (name) {
+        return { name, labels: labelData };
+      }
     }
```
So one remaining problem now is if you set a legend format, we preserve the labels. So getFieldDisplayText does not know that a legend format has been used and so you end up with legend format result and labels 
<img width="1219" alt="Screen Shot 2020-04-30 at 12 00 34" src="https://user-images.githubusercontent.com/10999/80698874-5d89bc00-8adb-11ea-9222-d96d632898f4.png">

torkelo commented on packages/grafana-data/src/field/fieldOverrides.ts:
```diff
@@ -95,23 +97,22 @@ export function applyFieldOverrides(options: ApplyFieldOverrideOptions): DataFra
-
     const scopedVars: ScopedVars = {
-      __series: { text: 'Series', value: { name } },
+      __series: { text: 'Series', value: { name: getFrameDisplayTitle(frame, index) } }, // might be missing
```
this is a problem as this makes __series.name to equal labels of the first field, and the default getTitleTemplate generates titles like `${__series.name}${__field.labels}`  which means field title with duplicated labels

torkelo committed:
thinking

torkelo committed:
More advanced and smart default title generation

torkelo committed:
Another fix

torkelo committed:
Progress but dam this will be hard

torkelo committed:
Merge remote-tracking branch 'origin/master' into better-label-support

torkelo committed:
Reverting the time series Value field name change

torkelo committed:
revert revert going in circles

torkelo commented:
The biggest tricky question for me is what should field.name be for time series? Value or rendered labels? If Value it makes field name matchers almost unusable (in transforms and overrides) and we would need special label matchers and transform options.

ryantxu commented:
Yes the current transformations all pick by just name -- I think that needs to be `name+tags`, since that is usually the real unique identifier.

Rather than calling it "Select by name" and can just be "Select field" and we use this id

torkelo commented:
> Yes the current transformations all pick by just name -- I think that needs to be name+tags, since that is usually the real unique identifier.

But that makes the concept of a field name really strange as it not really anything we can use directly. 

Means every time we refer to fields in transforms and matchers we should use getFieldId? hm.. maybe

Organize Fields transform, has a rename input to rename fields. Should the rename clear the labels? if not it won't actually rename what is shown, only change part of the field display text.

ryantxu commented:
> Organize Fields transform, has a rename input to rename fields. Should the rename clear the labels? if not it won't actually rename what is shown, only change part of the field display text.

I *think* that should set the title property 

The biggest complication is when the title is set, how does it affect the "fieldId"?  is it an additional option?

The "organize fields" transformer needs a warning that it only works with the first frame -- otherwise things get too crazy

torkelo commented:
> I think that should set the title property 

Yea probably, that way rename won't impact override rules you have configured (but will make them a bit more confusing) 

is there a way we can persist/build fieldId in pre transform step, having to build the id every time we need feels inefficient but might be fine, thinking also we need a field.getTitle() function or something that we attach to fields (Similar to display & getLinks) so getFieldDisplayText becomes somewhat discoverable :)

ryantxu committed:
Merge remote-tracking branch 'origin/master' into better-label-support

ryantxu committed:
add a field state object

torkelo committed:
Use state title when converting back to legacy format

torkelo commented:
Updated the PR description & TODO, Added notice about the current breaking change this PR means. 

# Breaking change 
This is currently a breaking change in the way that this will change how series names are built so in some cases for example for prometheus and graphite the resulting series names will be slightly different. For example if a query only returns a single label name, or single series name. This is due to the smarter logic in generating series names based on all series eliminating duplicate info. But also due the added white space between metric name and labels. If a legend format or alias function is used it should be fine, as the change in series name is only for the automatic series name.

So this will change some some percentage of panels that have colors set using series name, series overrides, or column styles. Creating a migration for this is next to impossible as we need to know the full query result to know what the series name will be.

I do not know how to evalaute the impact of this breaking change, and how we can avoid it minimize it.

torkelo committed:
Merge remote-tracking branch 'origin/master' into better-label-support

torkelo committed:
Merge remote-tracking branch 'origin/master' into better-label-support

torkelo committed:
Improved the join (series to columsn) transformer

torkelo committed:
Got tests running again

torkelo committed:
Rewrite of seriesToColums that simplifies and fixing tests

torkelo committed:
Fixed the tricky problem of multiple time field when not used in join

torkelo committed:
Prometheus: Restoring prometheus formatting

torkelo committed:
Graphite: Disable Grafana's series naming

torkelo committed:
fixed imports

torkelo committed:
Fixed tests and made rename transform change title instead

torkelo committed:
Fixing more tests

ryantxu committed:
Merge remote-tracking branch 'origin/master' into better-label-support

ryantxu committed:
fix more tests

torkelo committed:
fixed import issue

torkelo committed:
Fixed more circular dependencies

torkelo committed:
Renamed to getFieldTitle

hugohaggmark commented on packages/grafana-data/src/transformations/transformers/labelsToFields.test.ts:
```diff
@@ -43,38 +43,6 @@ describe('Labels as Columns', () => {
     expect(result[0].fields).toEqual(expected);
   });
 
```
Are we not supporting the removed test case any longer?

hugohaggmark replied:
Ah ok I see now, that we don't. So this transformer will have the same issues as before `getFramesWithOnlyValueFields` function was introduced with the long column names?

hugohaggmark replied:
Never mind, I thought we had revert the whole field naming logic.

torkelo replied:
yea, this logic is no longer needed :) as now time series always have the field name "Value" and no longer the same as the frame name, this removed test was for the logic that handled frame.name === field.name and that logic is no longer needed

torkelo committed:
More rename

hugohaggmark commented on packages/grafana-data/src/transformations/transformers/seriesToColumns.test.ts:
```diff
@@ -44,38 +45,53 @@ describe('SeriesToColumns Transformer', () => {
         values: new ArrayVector([1000, 3000, 4000, 5000, 6000, 7000]),
         config: {},
-        labels: { origin: 'even,odd' },
+        labels: undefined,
```
So I understand this better why is this `undefined` and not `{ name: 'even,odd' }`?

torkelo replied:
on the join by field we do not add origin, as the "title / identity" on the column you join by (I think) should not change

hugohaggmark commented on packages/grafana-data/src/transformations/transformers/seriesToColumns.ts:
```diff
@@ -16,55 +16,64 @@ export const seriesToColumnsTransformer: DataTransformerInfo<SeriesToColumnsOpti
-    if (!keyDataFrames.length) {
-      // for now we only parse data frames with 2 fields
-      return data;
+    const keyFieldMatch = options.byField || 'Time';
```
Maybe this is better?
```suggestion
    const keyFieldMatch = options.byField ?? defaultOptions.byField;
```

torkelo replied:
ah, good catch!

hugohaggmark commented on packages/grafana-data/src/types/dataFrame.ts:
```diff
@@ -18,6 +17,10 @@ export enum FieldType {
   other = 'other', // Object, Array, etc
 }
 
+export interface FieldCalcs {
+  [key: string]: any;
+}
+
```
nit
```suggestion
export interface FieldCalcs extends Record<string, any> {}
```

torkelo replied:
nice!

hugohaggmark commented on packages/grafana-data/src/field/fieldState.ts:
```diff
@@ -0,0 +1,148 @@
+  return `Series (${index})`;
+}
+
+export function getFieldTitle(field: Field, frame?: DataFrame, allFrames?: DataFrame[]): string {
```
There's something with the signature of this function that is hard to reason about. The field passed here should be part of one or several frames, so you should be able to figure out the field in question with rowIndex and fieldIndex, but I'm certainly missing something.

hugohaggmark commented on packages/grafana-ui/src/components/TransformersUI/OrganizeFieldsTransformerEditor.tsx:
```diff
@@ -71,6 +72,11 @@ const OrganizeFieldsTransformerEditor: React.FC<OrganizeFieldsTransformerEditorP
   );
 
+  // Show warning that we only apply the first frame
+  if (input.length > 1) {
+    return <div>Organize fields only works with a sigle frame. Consider applying a join transformation first.</div>;
+  }
```
Awesome! ⭐

hugohaggmark commented on public/app/features/dashboard/state/runRequest.ts:
```diff
@@ -169,9 +169,9 @@ export function getProcessedDataFrames(results?: DataQueryResponseData[]): DataF
+    // clear out the cached info
     for (const field of dataFrame.fields) {
-      field.calcs = null;
+      field.state = undefined;
```
nit: using `undefined` signals that this prop doesn't exists imo, `null` just states this prop is not initialized.
```suggestion
      field.state = null;
```

torkelo replied:
So should the Field type then be:

state?: FieldState | null; 

then we have another case to think about, hate undefined! 

Another alternative is:
delete field.state;

torkelo replied:
made it nullable, and set to null when cleared, think it makes more sense!

torkelo committed:
Review feedback

torkelo committed:
Merge remote-tracking branch 'origin/master' into better-label-support

dprokop commented:
Created this as a followup: https://github.com/grafana/grafana/issues/24337

torkelo committed:
Merge remote-tracking branch 'origin/master' into better-label-support

torkelo committed:
Fix for showing field title in calculate field transformer

torkelo committed:
fieldOverride: Make it clear that state title after applying defaults & overrides

torkelo committed:
Fixed ts issue

dprokop commented on packages/grafana-ui/src/components/TransformersUI/OrganizeFieldsTransformerEditor.tsx:
```diff
@@ -71,6 +72,11 @@ const OrganizeFieldsTransformerEditor: React.FC<OrganizeFieldsTransformerEditorP
 
+  // Show warning that we only apply the first frame
+  if (input.length > 1) {
+    return <div>Organize fields only works with a sigle frame. Consider applying a join transformation first.</div>;
```
```suggestion
    return <div>Organize fields only works with a single frame. Consider applying a join transformation first.</div>;
```

dprokop committed:
Update packages/grafana-ui/src/components/TransformersUI/OrganizeFieldsTransformerEditor.tsx

aocenas commented:
@torkelo @ryantxu this seems like a pretty big chunk of code to be merged during beta. How much testing did this have? To be honest we may not have that much capacity to test it thoroughly right now across explore and relevant data sources.

torkelo commented:
> this seems like a pretty big chunk of code to be merged during beta. How much testing did this have? To be honest we may not have that much capacity to test it thoroughly right now across explore and relevant data sources.

I know, but it is needed to make transforms work better in 7.0 and since it's one of the main features we really need this or might as well push the release. 

I have tested it with prometheus, loki, influxdb, elasticsearch , and ryan & dominik have also tested.
